plutus-tx/src/PlutusTx/Bool.hs

```diff
@@ -38,6 +38,42 @@ infixr 2 ||
 not :: Bool -> Bool
 not a = if a then False else True
 
+{- Note [Short-circuit evaluation for `&&` and `||`]
+TLDR: as it stands, the behaviors of `&&` and `||` are inconsistent: they sometimes
+short-circuit, sometimes don't.
+
+The Plutus Tx compiler treats `if-then-else` and `case` specially, and ensures that
+the branches aren’t strictly evaluated. For example, `if b then exprT else exprF` is
+compiled to `(if b then (/\a -> exprT) (/\a -> exprF)) @(forall a. a)`. The type
+abstractions suspend the evaluation of `exprT` and `exprF`. In PIR this looks like
+`Bool_match b (/\dead -> exprT) (/\dead -> exprF) {all dead . dead}`.
+
+Other than these, everything else is generally strictly evaluated. This includes
+`&&` and `||`, which are supposed to evaluate both arguments strictly. There are two
+problems with this: (1) `&&` and `||` should be made to short-circuit, otherwise they
+are not very useful; (2) in fact, their behaviors are not consistent - sometimes they
+short-circuit, sometimes they don't!
+
+Why do they short-circuit some of the times, but not always? This is due to the effect
+of GHC's unconditional inlining (a.k.a "pre-inlining"). We turn off as many GHC
+transformations as possible, since Plutus Tx and Haskell have somewhat different
+semantics, and a transformation made by GHC may not be valid (i.e., may change the program
+semantics) from PlutusTx's point of view. But we do run GHC's pre-inliner due to reasons
+explained in Note [GHC.sm_pre_inline]. In fact, in the current GHC version, we cannot
+completely turn off GHC's pre-inliner even if we want to, since it also runs in GHC's
+desugarar as part of the "simple optimizer".
+
+If GHC's pre-inliner inlines `&&` in `exprL && exprR`, it produces
+`if exprL then exprR else False`, enabling short-circuit evaluation (the second
+argument of `&&` must not be strict for this to happen - see
+Note [Lazy patterns on function parameters]).
+
+On the other hand, if the pre-inliner does not inline `&&`, then it does not
+short-circuit. Usually, the pre-inliner inlines `&&` if it occurs once, and
+does not if it occurs multiple times, such as in a program like
+`\x y z -> x < 3 && y < 4 && z < 5`.
```

Comment:
Nice write-up, I agree it makes practical sense to want `&&` and `||` to have short-circuiting semantics.

How is `if .. then .. else` treated specially? I understand how `case`  is treated specially, because `case` is part of the language, but I haven't seen `if .. then .. else`. Is it a builtin? Does it make sense to define a builtin without strict evaluation of its arguments? In any case, maybe we could we treat `&&` and `||` in the same way?

Sorry for all the questions, I just want to make sure I understand the details here.

Reply:
I started working on adding tests for this, and I think that this doesn't work as expected.

If `&&` short-circuits, then `False && error` should behave the same as `case False of {False -> False} {_ -> error}`, right?

```
case False of
  False -> False
  _ -> error ()

||
PIR
||
v

let
  data Bool | Bool_match where
    True : Bool
    False : Bool
in
False

||
UPLC
||
v

program 1.1.0 (constr 1 [])

||
eval
||
v

(constr 1 [])
```

The `case` evaluates, as expected, to `False`, and the error branch isn't evaluated. However, the `&&` program:

```
False && error ()

||
PIR
||
v

let
  data Bool | Bool_match where
    True : Bool
    False : Bool
  data Unit | Unit_match where
    Unit : Unit
  !x : Bool = error {Bool}  -- note that we get this binding here; why??
in
False

||
UPLC
||
v

program 1.1.0 ((\x -> constr 1 []) error)

||
eval
||
v

error
```

As noted above, the branch we expect to get short-circuited gets bound to a variable `x`, which eventually forces its evaluation.

Reply:
> How is if .. then .. else treated specially? I understand how case is treated specially, because case is part of the language, but I haven't seen if .. then .. else. 

Very good question. I don't think we actually treat `if-then-else` differently (at least not any more), since it is nothing but `case` on `Bool`. I'll update the Note.

Reply:
> How is `if .. then .. else` treated specially? I understand how `case` is treated specially, because `case` is part of the language, but I haven't seen `if .. then .. else`. Is it a builtin?

I was wondering about this myself, and it turns out to be surprisingly complicated.  We do have a built-in `ifThenElse` function, but that doesn't always get used: sometimes the standard Church lambda-calculus encoding is used instead.  For example, if you have 
```
fff :: Integer -> Integer -> Integer -> Integer
fff a b c = if a == 0 then b else c
```
then it gets compiled to
```
(lam a (lam b (lam c
 (force
  [
   [
    [(force (builtin ifThenElse)) [ [ (builtin equalsInteger) (con integer 0) ] a ] ]
    (delay b)
   ]
  (delay c)
  ]
))))
```

but
```
ggg :: Bool -> Integer -> Integer -> Integer
ggg a b c = if a then b else c
```
gets compiled to
```
(lam a (lam b (lam c (force (case a (delay b) (delay c))))))
```

The Haskell `if ... then ... else ...` gets desugared to `case ... of True -> ... False -> ...` in [GHC Core](https://downloads.haskell.org/ghc/9.0.1/docs/html/libraries/ghc-9.0.1/GHC-Core.html), so the plugin never sees it.  I think that what's happening is that everything initially gets converted into PIR using the Church version and then that sometimes gets converted into the form using `ifThenElse` when the PIR is simplified.  I'm having some difficulty working out  where that happens though.

Reply:
Man this is a mess. Perhaps worth reflecting this discussion in the Note itself.

Reply:
Yes, i agree it's a mess. We have to trace where this exactly happens.